src/h5cpp/datatype/enum.hpp

```diff
@@ -0,0 +1,214 @@
+//
+// (c) Copyright 2017 DESY,ESS
+//
+// This file is part of h5cpp.
+//
+// This library is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License as published
+// by the Free Software Foundation; either version 2.1 of the License, or
+// (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty ofMERCHANTABILITY
+// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+// License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this library; if not, write to the
+// Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor
+// Boston, MA  02110-1301 USA
+// ===========================================================================
+//
+// Authors:
+//   Eugen Wintersberger <eugen.wintersberger@desy.de>
+//   Martin Shetty <martin.shetty@esss.se>
+// Created on: May 14, 2018
+//
+#pragma once
+
+#include <h5cpp/datatype/datatype.hpp>
+#include <h5cpp/datatype/factory.hpp>
+#include <h5cpp/error/error.hpp>
+#include <sstream>
+
+namespace hdf5
+{
+namespace datatype
+{
+
+//!
+//! \brief compound data type
+//!
+//! An enum data type which can be used to save enum values along
+//! with metadata describing their meaning.
+//!
+class DLL_EXPORT Enum : public Datatype
+{
+ public:
+  //!
+  //! \brief default constructor
+  //!
+  Enum() = default;
+
+  //!
+  //! \brief construct from handle
+  //!
+  Enum(ObjectHandle&& handle);
+
+  //!
+  //! \brief construct from generic datatype
+  //!
+  Enum(const Datatype& type);
+
+  //!
+  //! \brief named constructor (with underlying type deduction)
+  //!
+  template<typename T>
+  static Enum create(const T&);
+
+  //!
+  //! \brief insert enum value definition (with underlying type deduction)
+  //!
+  template<typename T>
+  void insert(const std::string& name, const T& data) const;
+
+  //!
+  //! \brief return the total number of enum values
+  //!
+  size_t number_of_values() const;
+
+  //!
+  //! \brief get value name
+  //!
+  std::string name(size_t index) const;
+
+  // TODO: Cannot get name from value
+  // TODO: Discuss H5Tenum_nameof with HDF group, no way of obtaining string size
```

Comment:
Should these TODOs be left here? Better to make github tickets than leave TODOs in the code I think.